File: tray_detector/fusion.py
"""
fusion.py — Fusi Metode & Hierarki Eksekusi

Menggabungkan hasil dari Metode A, B, C menggunakan hierarki prioritas
dan weighted average untuk menghasilkan D_tray_cm final.

Hierarki:
  1. B + C keduanya valid → weighted average (B primary, C validasi)
  2. B saja valid → gunakan B
  3. C saja valid → gunakan C
  4. A saja → gunakan A (last resort)
  5. Tidak ada → INSUFFICIENT_DATA
"""


def fuse_results(result_a, result_b, result_c):
    """
    Fusi hasil dari ketiga metode sesuai hierarki prioritas.

    Args:
        result_a: dict dari method_a atau None
        result_b: dict dari method_b atau None
        result_c: dict dari method_c atau None

    Returns:
        dict: output schema final dengan semua field
    """
    b_valid = (result_b is not None and
               result_b.get("D_tray_cm") is not None and
               result_b.get("confidence", 0) > 0)

    c_valid = (result_c is not None and
               result_c.get("D_tray_cm") is not None and
               result_c.get("confidence", 0) > 0)

    a_valid = (result_a is not None and
               result_a.get("D_tray_cm") is not None and
               result_a.get("confidence", 0) > 0)

    # ── Prioritas 1: B + C keduanya valid ────────────────────────────────
    if b_valid and c_valid:
        D_b = result_b["D_tray_cm"]
        D_c = result_c["D_tray_cm"]
        conf_b = result_b["confidence"]
        conf_c = result_c["confidence"]

        D_fused = (D_b * conf_b + D_c * conf_c) / (conf_b + conf_c)
        conf_fused = min((conf_b + conf_c) / 2.0 * 1.1, 1.0)  # bonus fusi

        return _build_output(
            D_tray_cm=round(D_fused, 2),
            method_used="B+C",
            confidence=round(conf_fused, 3),
            status=result_b.get("status", "OK"),
            result_b=result_b,
            notes=None,
        )

    # ── Prioritas 2: B saja ──────────────────────────────────────────────
    if b_valid:
        return _build_output(
            D_tray_cm=result_b["D_tray_cm"],
            method_used="B",
            confidence=result_b["confidence"],
            status=result_b.get("status", "OK"),
            result_b=result_b,
            notes=result_b.get("notes"),
        )

    # ── Prioritas 3: C saja ──────────────────────────────────────────────
    if c_valid:
        return _build_output(
            D_tray_cm=result_c["D_tray_cm"],
            method_used="C",
            confidence=result_c["confidence"],
            status=result_c.get("status", "OK"),
            result_b=None,
            notes=result_c.get("notes"),
        )

    # ── Prioritas 4: A saja (last resort) ────────────────────────────────
    if a_valid:
        return _build_output(
            D_tray_cm=result_a["D_tray_cm"],
            method_used="A",
            confidence=result_a["confidence"],
            status=result_a.get("status", "OK"),
            result_b=None,
            notes=result_a.get("notes"),
        )

    # ── Gagal: tidak ada data valid ──────────────────────────────────────
    return _build_output(
        D_tray_cm=None,
        method_used="NONE",
        confidence=0.0,
        status="INSUFFICIENT_DATA",
        result_b=result_b,
        notes="Semua metode gagal — tidak ada data valid",
    )


def _build_output(D_tray_cm, method_used, confidence, status,
                   result_b=None, notes=None):
    """Bangun output dict sesuai schema dokumen."""
    return {
        "D_tray_cm": D_tray_cm,
        "method_used": method_used,
        "confidence": confidence,
        "status": status,
        "D_left_cm": result_b["D_left_cm"] if result_b else None,
        "D_right_cm": result_b["D_right_cm"] if result_b else None,
        "lines_left": result_b["lines_left"] if result_b else 0,
        "lines_right": result_b["lines_right"] if result_b else 0,
        "notes": notes,
    }

File: tray_detection_C/src/fusion.cpp
/**
 * fusion.cpp — Fusi hierarki metode A, B, C
 */

#include "fusion.h"
#include <cmath>
#include <algorithm>

namespace tray {

FusedResult fuse_results(
    const std::optional<MethodAResult>& result_a,
    const MethodBResult& result_b,
    const std::optional<MethodCResult>& result_c) {

    bool b_valid = result_b.valid && result_b.confidence > 0;
    bool c_valid = result_c.has_value() && result_c->valid
                   && result_c->confidence > 0;
    bool a_valid = result_a.has_value() && result_a->confidence > 0;

    FusedResult out;
    out.detail_b = result_b;
    if (result_c) out.detail_c = *result_c;

    // ── Priority 1: B + C ───────────────────────────────────────────────
    if (b_valid && c_valid) {
        double D_b = result_b.D_tray_cm;
        double D_c = result_c->D_tray_cm;
        double cb = result_b.confidence;
        double cc = result_c->confidence;

        out.D_tray_cm   = std::round((D_b * cb + D_c * cc) / (cb + cc) * 100.0) / 100.0;
        out.method_used = "B+C";
        out.confidence  = std::round(std::min((cb + cc) / 2.0 * 1.1, 1.0) * 1000.0) / 1000.0;
        out.status      = result_b.status;
        out.D_left_cm   = result_b.D_left_cm;
        out.D_right_cm  = result_b.D_right_cm;
        out.lines_left  = result_b.lines_left;
        out.lines_right = result_b.lines_right;
        return out;
    }

    // ── Priority 2: B only ──────────────────────────────────────────────
    if (b_valid) {
        out.D_tray_cm   = result_b.D_tray_cm;
        out.method_used = "B";
        out.confidence  = result_b.confidence;
        out.status      = result_b.status;
        out.D_left_cm   = result_b.D_left_cm;
        out.D_right_cm  = result_b.D_right_cm;
        out.lines_left  = result_b.lines_left;
        out.lines_right = result_b.lines_right;
        out.notes       = result_b.notes;
        return out;
    }

    // ── Priority 3: C only ──────────────────────────────────────────────
    if (c_valid) {
        out.D_tray_cm   = result_c->D_tray_cm;
        out.method_used = "C";
        out.confidence  = result_c->confidence;
        out.status      = result_c->status;
        out.notes       = result_c->notes;
        return out;
    }

    // ── Priority 4: A only ──────────────────────────────────────────────
    if (a_valid) {
        out.D_tray_cm   = result_a->D_tray_cm;
        out.method_used = "A";
        out.confidence  = result_a->confidence;
        out.status      = result_a->status;
        out.notes       = result_a->notes;
        return out;
    }

    // ── Nothing valid ───────────────────────────────────────────────────
    out.method_used = "NONE";
    out.status = "INSUFFICIENT_DATA";
    out.notes  = "Semua metode gagal";
    return out;
}


FusedResult build_single_result(
    const std::string& method,
    const std::optional<MethodAResult>& result_a,
    const MethodBResult& result_b,
    const std::optional<MethodCResult>& result_c) {

    FusedResult out;
    out.detail_b = result_b;
    if (result_c) out.detail_c = *result_c;

    if (method == "B" && result_b.valid) {
        out.D_tray_cm   = result_b.D_tray_cm;
        out.method_used = "B";
        out.confidence  = result_b.confidence;
        out.status      = result_b.status;
        out.D_left_cm   = result_b.D_left_cm;
        out.D_right_cm  = result_b.D_right_cm;
        out.lines_left  = result_b.lines_left;
        out.lines_right = result_b.lines_right;
        out.notes       = result_b.notes;
    } else if (method == "C" && result_c && result_c->valid) {
        out.D_tray_cm   = result_c->D_tray_cm;
        out.method_used = "C";
        out.confidence  = result_c->confidence;
        out.status      = result_c->status;
        out.notes       = result_c->notes;
    } else if (method == "A" && result_a && result_a->confidence > 0) {
        out.D_tray_cm   = result_a->D_tray_cm;
        out.method_used = "A";
        out.confidence  = result_a->confidence;
        out.status      = result_a->status;
        out.notes       = result_a->notes;
    } else {
        out.method_used = method;
        out.status = "INSUFFICIENT_DATA";
        out.notes = "Metode " + method + " gagal";
    }

    return out;
}

} // namespace tray
